onnxruntime/test/providers/cpu/tensor/transpose_test.cc

```diff
@@ -612,5 +613,16 @@ TEST(TransposeOpTest, Transpose0312) {
 }
 #endif
 
+// Test crafted with specific shape based on bug reported in GH issue : 7316
+TEST(TransposeOpTest, FloatDataPerm0213) {
+  std::vector<int64_t> input_shape({1, 1, 128, 1});
+  std::vector<float> input_vals(128, 1);
+
+  std::vector<int64_t> expected_shape({1, 128, 1, 1});
+
+  std::vector<int64_t> perm = {0, 2, 1, 3};
```

Comment:
For this test case, the input tensor can be shaped to {1, 128} before transpose. Can you improve the code below for it? https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/tensor/transpose.cc#L115-L156

Reply:
I crafted this simple test case to "mock" the scenario happening in the failing model referenced in the GH issue and trigger the code path that containing the bug. It is true that we can add more optimizations that can handle this particular test case -in fact for this specific test an optimization can be that we don't really need to do any transposing at all - it it is essentially just a buffer copy and a reshape. But adding more optimizations is not the scope of this PR. In the failing model, there is a 5-D tensor where two dims are flattened and the 4-D tensor  passes the check in CanDoTranspose4D()(). I just created a mock test that will go in the buggy code path.

Reply:
Removed this simplistic test case and added a better one